pkg/generic/binarypb_codec.go

```diff
@@ -0,0 +1,146 @@
+		}
+		return nil
+	}
+
+	methodName := msg.RPCInfo().Invocation().MethodName()
+	if methodName == "" {
+		return errors.New("empty methodName in protobuf Marshal")
+	}
+	// 3. encode metainfo
+	// 3.1 magic && msgType
+	if err := codec.WriteUint32(codec.ProtobufV1Magic+uint32(msg.MessageType()), out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write meta info failed: %s", err.Error()))
+	}
+	// 3.2 methodName
+	if _, err := codec.WriteString(methodName, out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write method name failed: %s", err.Error()))
+	}
+	// 3.3 seqID
+	if err := codec.WriteUint32(uint32(msg.RPCInfo().Invocation().SeqID()), out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write seqID failed: %s", err.Error()))
+	}
+
+	var transBuff []byte
+	var ok bool
+	if msg.RPCRole() == remote.Server {
+		gResult := data.(*Result)
+		transBinary := gResult.Success
+		if transBuff, ok = transBinary.(binaryPbReqType); !ok {
+			return perrors.NewProtocolErrorWithMsg("invalid marshal result in rawProtobufBinaryCodec: must be []byte")
+		}
+	} else {
+		gArg := data.(*Args)
+		transBinary := gArg.Request
+		if transBuff, ok = transBinary.(binaryPbReqType); !ok {
+			return perrors.NewProtocolErrorWithMsg("invalid marshal request in rawProtobufBinaryCodec: must be []byte")
+		}
+	}
+	out.WriteBinary(transBuff)
+	return nil
+}
+
+func (c *binaryProtobufCodec) Unmarshal(ctx context.Context, msg remote.Message, in remote.ByteBuffer) error {
+	payloadLen := msg.PayloadLen()
+	magicAndMsgType, err := codec.ReadUint32(in)
+	if err != nil {
+		return err
+	}
+	if magicAndMsgType&codec.MagicMask != codec.ProtobufV1Magic {
+		return perrors.NewProtocolErrorWithType(perrors.BadVersion, "Bad version in protobuf Unmarshal")
+	}
+	msgType := magicAndMsgType & codec.FrontMask
+	if err := codec.UpdateMsgType(msgType, msg); err != nil {
+		return err
+	}
+
+	methodName, methodFieldLen, err := codec.ReadString(in)
+	if err != nil {
+		return perrors.NewProtocolErrorWithErrMsg(err, fmt.Sprintf("protobuf unmarshal, read method name failed: %s", err.Error()))
+	}
+	if err = codec.SetOrCheckMethodName(methodName, msg); err != nil && msgType != uint32(remote.Exception) {
+		return err
+	}
+	seqID, err := codec.ReadUint32(in)
+	if err != nil {
+		return perrors.NewProtocolErrorWithErrMsg(err, fmt.Sprintf("protobuf unmarshal, read seqID failed: %s", err.Error()))
+	}
+	if err = codec.SetOrCheckSeqID(int32(seqID), msg); err != nil && msgType != uint32(remote.Exception) {
+		return err
+	}
+	actualMsgLen := payloadLen - metaInfoFixLen - methodFieldLen
+	transBuff, err := in.Next(actualMsgLen)
+	if err != nil {
+		return perrors.NewProtocolErrorWithErrMsg(err, fmt.Sprintf("protobuf unmarshal, read message buffer failed: %s", err.Error()))
+	}
```

Comment:
Binary generic is just for forwarding the whole data including metainfo, so transBuff length should be payloadLen.